### PR TITLE
Utility for comparing expressions

### LINKED ISSUE #1974: Add a function that can robustly compare expressions.
## Summary
This will be useful for testing.

A fairly easy way to implement this would be to "serialize" an expression into a list of nodes or node types.

Current tests that should use this:
  - pyomo/core/tests/unit/test_numeric_expr.py: `decompose_linear_term_wrapper` and `decompose_term_wrapper`

### PR DESCRIPTION
**Pull request by michaelbynum**
## Summary/Motivation:
This PR adds utilities for comparing expressions.

This PR depends on #1954.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/expr/compare.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
+from .visitor import StreamBasedExpressionVisitor
```
Comment (mrmundt) [inline pyomo/core/expr/compare.py:1]: This file is missing the copyright.

File: pyomo/core/tests/unit/test_compare.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
+from pyomo.common import unittest
```
Comment (mrmundt) [inline pyomo/core/tests/unit/test_compare.py:1]: This file is missing the copyright.

**Review by mrmundt — DISMISSED** [review]

File: pyomo/core/expr/compare.py (hunk 1 above)
Comment (michaelbynum) [reply pyomo/core/expr/compare.py:10]: Fixed

File: pyomo/core/tests/unit/test_compare.py (hunk 2 above)
Comment (michaelbynum) [reply pyomo/core/tests/unit/test_compare.py:10]: Fixed

File: pyomo/core/expr/compare.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+handler[NPV_AbsExpression] = handle_unary_expression
+handler[RangedExpression] = handle_expression
+handler[InequalityExpression] = handle_expression
+handler[EqualityExpression] = handle_expression
+
+
+class PrefixVisitor(StreamBasedExpressionVisitor):
+    def __init__(self):
+        super().__init__()
+        self._result = list()
```
Comment (jsiirola) [inline pyomo/core/expr/compare.py:77]: I would initialize this to `None`: `initializeWalker` should take care of creating the list. (the one you create here is just thrown away)

File: pyomo/core/expr/compare.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            return tuple(), None
+
+        if node.is_expression_type():
+            return handler[ntype](node, self._result), None
+        else:
+            self._result.append(node)
+            return tuple(), None
+
+    def finalizeResult(self, result):
+        return self._result
```
Comment (jsiirola) [inline pyomo/core/expr/compare.py:109]: I would reset `_result` to `None` here:
```python
ans = self._result
self._result = None
return ans
```

File: pyomo/core/expr/compare.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            return False, self._result
+
+        if expr.is_expression_type():
+            if etype is LinearExpression:
+                handle_linear_expression(expr, self._result)
+                return False, self._result
+            return True, None
+
+        self._result.append(expr)
+        return False, self._result
```
Comment (jsiirola) [inline pyomo/core/expr/compare.py:94]: There is no need to replicate the logic of `enterNode` here.

**Review by jsiirola — APPROVED** [review]
This looks pretty good.  A couple minor comments that I am putting in for reference and will go ahead and directly fix on the PR momentarily.

**Review by blnicho — APPROVED** [review]

**Pull request merged**